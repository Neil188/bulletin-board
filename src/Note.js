import React, { Component } from 'react';
import Draggable from 'react-draggable';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaTrash from 'react-icons/lib/fa/trash';
import FaFloppyO from 'react-icons/lib/fa/floppy-o';

class Note extends Component {
    constructor(props) {
        super(props);
        this.state = {
            editing: false,
        };
        this.edit = this.edit.bind(this);
        this.randomBetween = this.randomBetween.bind(this);
        this.remove = this.remove.bind(this);
        this.save = this.save.bind(this);
        this.renderForm = this.renderForm.bind(this);
        this.renderDisplay = this.renderDisplay.bind(this);
    }

    componentWillMount() {
        this.style = {
            right: this.randomBetween(0, window.innerWidth - 150, 'px'),
            top: this.randomBetween(0, window.innerHeight - 150, 'px'),
            transform: `rotate( ${this.randomBetween(-25, 25, 'deg')} )`,
        };
    }

    shouldComponentUpdate( nextProps, nextState ) {
        return (
            this.props.children !== nextProps.children ||
            this.state !== nextState
        );
    }

    componentDidUpdate() {
        let textArea;
        if (this.state.editing) {
            textArea = this.newText;
            textArea.focus();
            textArea.select();
        }
    }

    edit() {
        this.setState({
            editing: true,
        });
    }

    randomBetween(x, y, s) {
        return x + Math.ceil( Math.random() * (y-x) ) + s;
    }

    remove() {
        this.props.onRemove(this.props.index);
    }

    save(e) {
        e.preventDefault();
        this.props.onChange(this.newText.value, this.props.index);
        this.setState({
            editing: false,
        });
    }

    renderForm() {
        return (
            <Draggable disabled='true'>
                <div>
                    <div className='note' style={this.style}>
                        <form onSubmit={this.save}>
                            <textarea
                                ref={ input => {this.newText= input;} }
                                defaultValue={this.props.children}
                                maxLength={40}
                                cols={18}
                                rows={3}
                            />
                            <button id='save'>
                                <FaFloppyO />
                            </button>
                        </form>
                    </div>
                </div>
            </Draggable>
        );
    }

    renderDisplay() {
        return (
            <Draggable>
                <div>
                    <div className='note' style={this.style}>
                        <p>{this.props.children}</p>
                        <span>
                            <button
                                onClick={this.edit}
                                id='edit'
                            >
                                <FaPencil />
                            </button>
                            <button
                                onClick={this.remove}
                                id='remove'
                            >
                                <FaTrash />
                            </button>
                        </span>
                    </div>
                </div>
            </Draggable>
        );
    }

    render() {
        return this.state.editing ?
            this.renderForm()
            : this.renderDisplay();
    }
}

export default Note;