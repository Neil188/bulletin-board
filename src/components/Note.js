import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Draggable from 'react-draggable';
import FaPencil from 'react-icons/lib/fa/pencil';
import FaTrash from 'react-icons/lib/fa/trash';
import FaFloppyO from 'react-icons/lib/fa/floppy-o';
import { calcRandomBetween } from '../utils/random';

class Note extends Component {

    state = {
        editing: false,
    };

    componentWillMount() {
        this.style = {
            right: calcRandomBetween(0, window.innerWidth - 150, 'px'),
            top: calcRandomBetween(0, window.innerHeight - 150, 'px'),
            transform: `rotate( ${calcRandomBetween(-25, 25, 'deg')} )`,
        };
    }

    shouldComponentUpdate( nextProps, nextState ) {
        return (
            this.props.children !== nextProps.children ||
            this.state !== nextState
        );
    }

    // componentDidUpdate() {
    //     let textArea;
    //     if (this.state.editing) {
    //         textArea = this.newText;
    //         // textArea.focus();
    //         // textArea.select();
    //     }
    // }

    edit = () => {
        this.setState(() => ({
            editing: true,
        }));
    }

    remove = () =>
        this.props.onRemove(this.props.index);

    save = (e) => {
        e.preventDefault();
        this.props.onChange(this.newText.value, this.props.index);
        this.setState(() => ({
            editing: false,
        }));
    }

    renderForm = () => (
        <Draggable disabled>
            <div>
                <div className='note' style={this.style}>
                    <form onSubmit={this.save}>
                        <textarea
                            ref={ input => {this.newText= input;} }
                            defaultValue={this.props.children}
                            maxLength={40}
                            rows={4}
                        />
                        <button
                            className='save'
                            aria-label='save'
                        >
                            <FaFloppyO />
                        </button>
                    </form>
                </div>
            </div>
        </Draggable>
    )

    renderDisplay = () =>(
        <Draggable>
            <div>
                <div className='note' style={this.style}>
                    <p>{this.props.children}</p>
                    <span>
                        <button
                            onClick={this.edit}
                            className='edit'
                            aria-label='edit'
                        >
                            <FaPencil />
                        </button>
                        <button
                            onClick={this.remove}
                            className='remove'
                            aria-label='remove'
                        >
                            <FaTrash />
                        </button>
                    </span>
                </div>
            </div>
        </Draggable>
    )

    render() {
        return this.state.editing ?
            this.renderForm()
            : this.renderDisplay();
    }
}

Note.propTypes = {
    index: PropTypes.number.isRequired,
    children: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
};

export default Note;
