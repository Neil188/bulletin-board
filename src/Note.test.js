import React from 'react';
import { create } from 'react-test-renderer';
import ReactDOM from 'react-dom';
import Note from './Note';

describe('Note snapshots', () => {
    test('Note snapshot test', () => {
        const component = create(
            <Note index={0} onChange={() => null} onRemove={() => null} test>
                Test
            </Note>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Check for render failures', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <Note index={0} onChange={() => null} onRemove={() => null} test>
                Test
            </Note>
            , div);
    });
});