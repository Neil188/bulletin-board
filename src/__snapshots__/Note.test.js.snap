// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note snapshots Note snapshot test 1`] = `
<div
  className="react-draggable"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "touchAction": "none",
      "transform": "translate(0px,0px)",
    }
  }
  transform={null}
>
  <div
    className="note"
    style={
      Object {
        "right": 0,
        "top": 0,
        "transform": "rotate( 0 )",
      }
    }
  >
    <p>
      Test
    </p>
    <span>
      <button
        aria-label="edit"
        className="edit"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m11.1 34.3l2-2-5.2-5.3-2 2v2.4h2.8v2.9h2.4z m11.7-20.7q0-0.5-0.5-0.5-0.2 0-0.4 0.1l-12.1 12.1q-0.1 0.2-0.1 0.4 0 0.5 0.4 0.5 0.3 0 0.4-0.2l12.1-12.1q0.2-0.1 0.2-0.3z m-1.2-4.3l9.3 9.3-18.6 18.5h-9.3v-9.2z m15.2 2.1q0 1.2-0.8 2l-3.7 3.7-9.3-9.2 3.7-3.7q0.8-0.9 2-0.9 1.2 0 2 0.9l5.3 5.2q0.8 0.9 0.8 2z"
            />
          </g>
        </svg>
      </button>
      <button
        aria-label="remove"
        className="remove"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m15.9 30.7v-15.7q0-0.3-0.2-0.5t-0.5-0.2h-1.4q-0.3 0-0.5 0.2t-0.2 0.5v15.7q0 0.3 0.2 0.5t0.5 0.2h1.4q0.3 0 0.5-0.2t0.2-0.5z m5.7 0v-15.7q0-0.3-0.2-0.5t-0.5-0.2h-1.4q-0.3 0-0.5 0.2t-0.2 0.5v15.7q0 0.3 0.2 0.5t0.5 0.2h1.4q0.3 0 0.5-0.2t0.2-0.5z m5.8 0v-15.7q0-0.3-0.2-0.5t-0.6-0.2h-1.4q-0.3 0-0.5 0.2t-0.2 0.5v15.7q0 0.3 0.2 0.5t0.5 0.2h1.4q0.4 0 0.6-0.2t0.2-0.5z m-12.2-22.1h10l-1.1-2.6q-0.1-0.2-0.3-0.3h-7.1q-0.2 0.1-0.4 0.3z m20.7 0.7v1.4q0 0.3-0.2 0.5t-0.5 0.2h-2.1v21.2q0 1.8-1.1 3.2t-2.5 1.3h-18.6q-1.4 0-2.5-1.3t-1-3.1v-21.3h-2.2q-0.3 0-0.5-0.2t-0.2-0.5v-1.4q0-0.3 0.2-0.5t0.5-0.2h6.9l1.6-3.8q0.3-0.8 1.2-1.4t1.7-0.5h7.2q0.9 0 1.8 0.5t1.2 1.4l1.5 3.8h6.9q0.3 0 0.5 0.2t0.2 0.5z"
            />
          </g>
        </svg>
      </button>
    </span>
  </div>
</div>
`;

exports[`Note snapshots Note snapshot test 2`] = `
<div
  className="react-draggable"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "touchAction": "none",
      "transform": "translate(0px,0px)",
    }
  }
  transform={null}
>
  <div
    className="note"
    style={
      Object {
        "right": 0,
        "top": 0,
        "transform": "rotate( 0 )",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <textarea
        defaultValue="Test"
        maxLength={40}
        rows={4}
      />
      <button
        aria-label="save"
        className="save"
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": undefined,
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m11.6 34.3h17.1v-8.6h-17.1v8.6z m20 0h2.8v-20q0-0.3-0.2-0.9t-0.4-0.7l-6.3-6.3q-0.2-0.2-0.8-0.5t-0.8-0.2v9.3q0 0.9-0.7 1.5t-1.5 0.6h-12.8q-0.9 0-1.6-0.6t-0.6-1.5v-9.3h-2.8v28.6h2.8v-9.3q0-0.9 0.6-1.5t1.6-0.6h18.5q0.9 0 1.5 0.6t0.7 1.5v9.3z m-8.6-20.7v-7.2q0-0.3-0.2-0.5t-0.5-0.2h-4.3q-0.3 0-0.5 0.2t-0.2 0.5v7.2q0 0.3 0.2 0.5t0.5 0.2h4.3q0.3 0 0.5-0.2t0.2-0.5z m14.3 0.7v20.7q0 0.9-0.6 1.5t-1.6 0.6h-30q-0.8 0-1.5-0.6t-0.6-1.5v-30q0-0.9 0.6-1.5t1.5-0.6h20.8q0.9 0 1.9 0.4t1.7 1.1l6.3 6.2q0.6 0.6 1 1.7t0.5 2z"
            />
          </g>
        </svg>
      </button>
    </form>
  </div>
</div>
`;
